### ISSUE TITLE
Handling of kernelspecs

### ISSUE THREAD

**Issue by n-riesco** [nteract/hydrogen#272]
I'm moving the discussion about kernelspecs from #271 to this issue.

@slavaGanzin @rgbkrk Please, could you clarify your position? Are you supporting making the installation of Jupyter a requirement to use Hydrogen?

The latest Ubuntu 16.04 LTS comes with `IPython v2.4.1`. This version of IPython doesn't support the command `ipython kernelspec`. This is very unfortunate and it's raising the hurdle users have to hop in order to use Hydrogen.

@rgbkrk Would any of the `nteract` libraries be ready to help us here? If not, I would consider contributing to them so that we can solve this unfortunate situation sooner than later.

**Comment by slavaGanzin** [comment]
I don't care about Ubuntu problems, but as I said before:
 if ```jupyter/ipython kernelspec``` fails hydrogen should use kernelspec mappings that can be set in hydrogen settings. That means jupyter wouldn't be a requirement for hydrogen.

if we can get kernelspecs from ipython/jupyter, than that mappings should be merged internally (without showing them in settings) with kernelspecs from settings. And so if kernel was uninstalled from system it wouldn't be listed in kernel list (as now).
In this way you can manage only remote kernels (or specific ones) in settings.

**Comment by rgbkrk** [comment]
[`kernelspecs`](https://github.com/nteract/kernelspecs) is the library for querying kernel specs, though it only relies on `jupyter-paths` at the moment. We could introduce a way to check for `ipython-paths` as well as old IPython. Based on the fact that we "require" a kernel to be installed, even with `jupyter`, we likely need a hokey default kernel to be provided back from kernelspecs.

**Comment by slavaGanzin** [comment]
> I don't care about Ubuntu problems
@n-riesco Life is ironic, even sarcastic - now I work on ubuntu 16.04 LTS (laughing allowed)

I read docs for ijavascript, jp-babel and installed env as described:
```sudo apt-get install nodejs-legacy npm ipython ipython-notebook libzmq3-dev```

and was discouraged with:

```
updateKernelSpecs: `jupyter kernelspec` failed Error: Command failed: /bin/sh -c jupyter kernelspec list --json --log-level=CRITICAL(…)
kernel-manager.coffee:71 updateKernelSpecs: `ipython kernelspec` failed Error: Command failed: /bin/sh -c ipython kernelspec list --json --log-level=CRITICAL(…)
```

after that I try to intall ```jupyter``` from pip and everything works like a magic. 

So here is a question why do you recomend to use constantly outdated ubuntu repositories if anyone can get fresh and shiny from pip?

p.s. I hate ubuntu and idea about using LTS on desktop, so I'll switch. But Ubuntu is fucking mainstream and Hydrogen users shouldn't suffer

**Comment by slavaGanzin** [comment]
And another issue: I sync my settings between work and home laptops. So I have specified kernels in my settings. Update kernels on work machine doesn't update my kernelspecs because I have ones from my home. That's the weirdest behaviour I ever expect.

I'll implement that this way: https://github.com/nteract/hydrogen/issues/271#issuecomment-220771030
Current behaviour is a mess

**Comment by n-riesco** [comment]
To get the discussion in this issue re-started. Here are the main questions I thought we need to answer:

1. How to locate kernel specs
    - Currently: use `jupyter kernelspec`
    - Proposed: use a javascript-only library

2. How to set the setting `kernelSpec`
    - Currently: merge user setting with output of `jupyter kernelspec`
    - Proposed: `kernelSpec` only stores user setting

3. How to set the setting `languageMappings`
    - Currently: set by user to help with kernel languages that don't match grammar languages
    - Proposed: remove `languageMappings` and use only `kernelMappings`

4. How to set the setting `kernelMappings`
    - Currently: command `hydrogen:select-kernel` only shows kernels that match grammar language
    - Proposed: command `hydrogen:select-kernel` shows all kernels regardless of whether they match the grammar language

5. How to start a kernel for a given file:
    - Currently:
      - use the connection file `./hydrogen/connection.json`
      - use the setting `kernelMappings`
      - match grammar and kernel languages
    - Proposed:
      - use the kernel spec `./hydrogen/kernel.json`

**Comment by lgeiger** [comment]
Just a question at the beginning:
Do we need the kernelSpec setting if we make `jupyter` a dependency? This would simplify the code a lot and would make bullet point 1 very easy to achieve plus it is easy to install with `pip/conda install jupyter` and `ipython kernel install` on every system and easier than editing json. For my understanding this setting is here for useres who have an old `ipython` version, or am I missing something?

1. I would support the move to [`kernelspecs`](https://github.com/nteract/kernelspecs).
2. If we need the kernelSpec setting, I would prefer the proposed option.
3. I'm a bit confused with these settings and I think I'm not the only one. I would appreciate a merge of them if possible, so we can directly map kernel language names to kernel display names (eg. `{'somestrangekernelname': 'Python 3'}`). @slavaGanzin Somewhere you explained how you use them, but I couldn't find the issue, would your workflow work with this option?
4. I prefer the current option, because it's easier for new users. But I understand why the proposed option can be useful. I think the best way is a new setting to give the advanced user a choice.

**Comment by n-riesco** [comment]
On 05/07/16 22:49, Lukas Geiger wrote:
> Just a question at the beginning:
> Do we need the kernelSpec setting if we make |jupyter| a dependency?

I'm against making Jupyter a dependency. I'd like Hydrogen to run out of the box (i.e. no need to install ZMQ or Jupyter; we're far from achieving this).

Otherwise, we won't stop the constant flow of issues reporting problems with the installation of ZMQ and Jupyter.

> This would simplify the code a lot and would make bullet point 1 very easy to achieve plus it is easy to install with |pip/conda install jupyter| and |ipython kernel install| on every system and easier than editing json. For my understanding this setting is here for useres who have an old |ipython| version, or am I missing something?

Ditto: we would have to deal with issues reported by users expecting Hydrogen to work with the IPython version in Ubuntu 16.04.

>  1. I would support the move to |kernelspecs| <https://github.com/nteract/kernelspecs>.

As explained above, I still believe in the merits of having a javascript-only solution.

>  2. If we need the kernelSpec setting, I would prefer the proposed option.

Me too.

>  3. I'm a bit confused with these settings and I think I'm not the only one. I would appreciate a merge of them if possible, so we can directly map kernel language names to kernel display names (eg. |{'somestrangekernelname': 'Python 3'}|). @slavaGanzin <https://github.com/slavaGanzin> Somewhere you explained how you use them, but I couldn't find the issue, would your workflow work with this option?

It was confusing because the code was using language names from too many sources (that sometimes were lowercased, but not always). This has been the source of many issue reports.

With PR #260, we reduced the sources of language names to 2:
- lowercased Atom grammar language names (used to identify what kernels can run a given grammar language)
- kernel display names (used to identify a kernel)

If we removed `languageMappings`, `hydrogen:select-kernel` won't list kernels that declare non-standard language names.

To workaround the removal of `languageMappings`, we change `hydrogen:select-kernel` to list all kernels, but this is very inconvenient for users like me with a large number of kernels.

>  4. I prefer the current option, because it's easier for new users. But I understand why the proposed option can be useful. I think the best way is a new setting to give the advanced user a choice.

I'm not against, the removal of `languageMappings`, but I'd like to make sure we provide the user the best experience we can:
- e.g. having a setting to configure whether `hydrogen:select-kernel` shows all the kernels
- e.g. having two commands `hydrogen:select-kernel` and `hydrogen:select-non-standard-kernel`
- ...

**Comment by slavaGanzin** [comment]
1. "javascript-only solution" is just hardcoded paths and it often fail.
Parsing ```jupyter kernelspecs``` should be moved to https://github.com/nteract/kernelspecs by design. And used by hydrogen as dependency.

2. ```Jupyter/ipython kernelspecs``` should be optional, but preferred way. Jupyter/ipython shouldn't be a dependency. But it should be recommended.

3. I think two options for selecting kernel is good choice.

4. languageMappings should be removed as redundant.

**Comment by n-riesco** [comment]
Closed via #351.
